presto-orc/src/main/java/com/facebook/presto/orc/reader/ApacheHiveTimestampDecoder.java

```diff
@@ -36,8 +37,30 @@ public static long decodeTimestamp(long seconds, long serializedNanos, DecodeTim
         if (value < 0 && nanos != 0) {
             value -= options.getUnitsPerSecond();
         }
-        // Truncate nanos to required units (millis / micros) and add to value
-        return value + (nanos / options.getNanosPerUnit());
+        return getValueWithNanos(value, nanos, options.getNanosPerUnit());
+    }
+
+    private static long getSecondsInRequiredUnits(long secondsWithBase, long unitsPerSecond)
```

Comment:
We want the milliseconds to use the regular multiplication. Detecting milliseconds on overflow is a breaking change and many queries would fail. We don't want that to change with this PR.

Reply:
Thanks, I have added additional check to detect overflow only when micro precision is enabled.